src/core/Do.Architecture/Architecture/ApplicationContext.cs

```diff
@@ -6,41 +6,26 @@ public class ApplicationContext
 
     public void Add<T>(T item) where T : notnull => _context[typeof(T)] = item;
 
+    public bool Has<T>() => _context.ContainsKey(typeof(T));
+
     public T Get<T>()
     {
         if (!Has<T>())
         {
-            throw BuildKeyNotFound<T>();
+            throw new KeyNotFoundException($"'{typeof(T).Name}' does not exist in context, " +
+                $"{BuildMessageDetails<T>()}"
+            );
         }
 
         return (T)_context[typeof(T)];
     }
 
-    public bool Has<T>() => _context.ContainsKey(typeof(T));
-
-    private KeyNotFoundException BuildKeyNotFound<T>()
-    {
-        var message = $"'{typeof(T).Name}' does not exist in context.";
-
-        if (_context.Count == 0)
-        {
-            message = message.Replace('.', ' ') + "because it is empty.";
+    private string BuildMessageDetails<T>() =>
+        !_context.Any()
+            ? "because it is empty."
+            : GetAssignableTypes<T>().Any()
+                ? $"did you mean: {string.Join(", ", GetAssignableTypes<T>().Select(t => $"'{t.Name}'"))}?"
+                : $"available types are: {string.Join(", ", _context.Keys.Select(t => $"'{t.Name}'"))}.";
```

Comment:
i think it's better to create a method `GetTypes` for `_context.Keys` to improve readability, and have a consistency with `GetAssignableTypes`